File: packages/python/examples/duplicate_detection.py
#!/usr/bin/env python3
"""Duplicate detection using LSH bands.

This example demonstrates:
- Batch signature generation for multiple vectors
- Using LSH bands for efficient candidate generation
- Finding near-duplicates in a collection

Run with: python python/examples/duplicate_detection.py
"""

from collections import defaultdict
from odin_sig import (
    cosine_from_hamming,
    hamming_distance_hex,
    normalize_vector,
    simhash_lsh_multi,
)


def main():
    print("=== Duplicate Detection with LSH ===\n")

    # Example vectors representing different documents
    # Vectors 0, 1, 2 are similar (duplicates)
    # Vectors 3, 4 are different
    vectors = [
        [1.0, 1.0, 1.0, 1.0],  # Doc 0
        [1.0, 0.95, 1.05, 1.0],  # Doc 1 (near-duplicate of 0)
        [0.98, 1.02, 1.0, 1.01],  # Doc 2 (near-duplicate of 0, 1)
        [0.0, 1.0, 0.0, 1.0],  # Doc 3 (different)
        [-1.0, -1.0, -1.0, -1.0],  # Doc 4 (opposite, different)
    ]

    print(f"Processing {len(vectors)} documents...\n")

    # Normalize and generate signatures
    signatures = [simhash_lsh_multi(normalize_vector(v)) for v in vectors]

    # Build band index for candidate generation
    # Map: (band_index, band_value) -> [doc_ids]
    band_index = defaultdict(list)

    for doc_id, sig in enumerate(signatures):
        # Use first family only for this example
        family = sig[0]

        for band_idx, band_value in enumerate(family.bands):
            key = (band_idx, band_value)
            band_index[key].append(doc_id)

    # Find candidate pairs (documents that share at least one band)
    candidates = set()

    for docs in band_index.values():
        if len(docs) > 1:
            # Multiple documents match this band
            for i in range(len(docs)):
                for j in range(i + 1, len(docs)):
                    pair = tuple(sorted([docs[i], docs[j]]))
                    candidates.add(pair)

    print(f"Found {len(candidates)} candidate pairs from band matching\n")

    # Verify candidates with full Hamming distance
    threshold = 0.85  # Cosine similarity threshold for duplicates
    duplicates = []

    for id1, id2 in candidates:
        sig1 = signatures[id1][0].signature
        sig2 = signatures[id2][0].signature

        hamming = hamming_distance_hex(sig1, sig2)
        similarity = cosine_from_hamming(hamming, 256)

        if similarity >= threshold:
            duplicates.append((id1, id2, similarity))

    # Sort by similarity (descending)
    duplicates.sort(key=lambda x: x[2], reverse=True)

    print(f"Detected duplicates (similarity >= {threshold}):\n")
    for id1, id2, sim in duplicates:
        print(f"  Doc {id1} <-> Doc {id2}: {sim:.4f}")

    print("\n✓ Duplicate detection complete!")
    print("\nKey insight:")
    print("  - Band matching reduces comparisons from O(n²) to O(n)")
    print("  - Only candidate pairs need full Hamming distance computation")
    print("  - Tune bands/bits ratio for precision/recall tradeoff")


if __name__ == "__main__":
    main()

File: packages/typescript/examples/duplicate_detection.ts
#!/usr/bin/env ts-node
/**
 * Duplicate detection using LSH bands.
 *
 * This example demonstrates:
 * - Batch signature generation for multiple vectors
 * - Using LSH bands for efficient candidate generation
 * - Finding near-duplicates in a collection
 *
 * Run with: npx ts-node typescript/examples/duplicate_detection.ts
 */

import {
  cosineFromHamming,
  hammingDistanceHex,
  normalizeVector,
  simhashLshMulti,
  type LSHFamily,
} from '../src/lsh';

function main() {
  console.log('=== Duplicate Detection with LSH ===\n');

  // Example vectors representing different documents
  // Vectors 0, 1, 2 are similar (duplicates)
  // Vectors 3, 4 are different
  const vectors = [
    [1.0, 1.0, 1.0, 1.0],      // Doc 0
    [1.0, 0.95, 1.05, 1.0],    // Doc 1 (near-duplicate of 0)
    [0.98, 1.02, 1.0, 1.01],   // Doc 2 (near-duplicate of 0, 1)
    [0.0, 1.0, 0.0, 1.0],      // Doc 3 (different)
    [-1.0, -1.0, -1.0, -1.0],  // Doc 4 (opposite, different)
  ];

  console.log(`Processing ${vectors.length} documents...\n`);

  // Normalize and generate signatures
  const signatures: LSHFamily[][] = vectors.map(v => 
    simhashLshMulti(normalizeVector(v))
  );

  // Build band index for candidate generation
  // Map: (band_index, band_value) -> [doc_ids]
  const bandIndex = new Map<string, number[]>();

  for (let docId = 0; docId < signatures.length; docId++) {
    // Use first family only for this example
    const family = signatures[docId][0];

    for (let bandIdx = 0; bandIdx < family.bands.length; bandIdx++) {
      const key = `${bandIdx}:${family.bands[bandIdx]}`;
      const docs = bandIndex.get(key) || [];
      docs.push(docId);
      bandIndex.set(key, docs);
    }
  }

  // Find candidate pairs (documents that share at least one band)
  const candidates = new Set<string>();

  for (const docs of bandIndex.values()) {
    if (docs.length > 1) {
      // Multiple documents match this band
      for (let i = 0; i < docs.length; i++) {
        for (let j = i + 1; j < docs.length; j++) {
          const pair = [docs[i], docs[j]].sort((a, b) => a - b).join(',');
          candidates.add(pair);
        }
      }
    }
  }

  console.log(`Found ${candidates.size} candidate pairs from band matching\n`);

  // Verify candidates with full Hamming distance
  const threshold = 0.85; // Cosine similarity threshold for duplicates
  const duplicates: [number, number, number][] = [];

  for (const pair of candidates) {
    const [id1, id2] = pair.split(',').map(Number);
    const sig1 = signatures[id1][0].signature;
    const sig2 = signatures[id2][0].signature;

    const hamming = hammingDistanceHex(sig1, sig2);
    const similarity = cosineFromHamming(hamming, 256);

    if (similarity >= threshold) {
      duplicates.push([id1, id2, similarity]);
    }
  }

  // Sort by similarity (descending)
  duplicates.sort((a, b) => b[2] - a[2]);

  console.log(`Detected duplicates (similarity >= ${threshold}):\n`);
  for (const [id1, id2, sim] of duplicates) {
    console.log(`  Doc ${id1} <-> Doc ${id2}: ${sim.toFixed(4)}`);
  }

  console.log('\n✓ Duplicate detection complete!');
  console.log('\nKey insight:');
  console.log('  - Band matching reduces comparisons from O(n²) to O(n)');
  console.log('  - Only candidate pairs need full Hamming distance computation');
  console.log('  - Tune bands/bits ratio for precision/recall tradeoff');
}

main();
